use spin controls for numerical input

feerrenrut opened the pull request:
Fix for #6099

For numerical input fields a wx SpinCtrl is now used. The min/max values
come from the nvda config schema validation (if available).

It would be nice to access these values programmatically rather than duplicate them, however I could not see a way to query configObj for this.

In terms of accessibility, NVDA outputs:

> Cursor blink rate (ms)  edit  100

When trying to type non numerics:

> Unacceptable Character: You can only type a number here.  help balloon

feerrenrut committed:
fix for #6099 use spin controls

For numerical input fields a wx SpinCtrl is now used. The min/max values
come from the nvda config schema validation (if available).

feerrenrut commented:
@jcsteh mind reviewing this when you get the chance?

jcsteh commented on source/gui/settingsDialogs.py:
```diff
@@ -517,8 +517,8 @@ def makeSettings(self, settingsSizer):
 		settingsSizer.Add(capPitchChangeLabel)
-		self.capPitchChangeEdit=wx.TextCtrl(self,wx.NewId())
-		self.capPitchChangeEdit.SetValue(str(config.conf["speech"][getSynth().name]["capPitchChange"]))
+		self.capPitchChangeEdit=wx.SpinCtrl(self, min=0, max=100,
```
The minimum here is -100, not 0. (The value is relative, not multiplicative.) Do these controls support negative values?

jcsteh reviewed (changes requested):
1. For those that don't have defined ranges in the config spec, they have ranges enforced by the `onOk` method. For example, maxLineLength has a min of 10 and a max of 250. These ranges should be used for the spin control.
2. Once these ranges are set correctly, I think it'd make sense to remove the validation from `onOk`, since it's already being done by the control.
3. The contents of the field isn't selected by default when it gets focus. This is a change from the existing behaviour and means that a user can't simply tab to the field and start typing to replace the previous value. I think this is potentially going to annoy users, so I suggest subclassing the control to make it select the contents of the field when it gets focus.

feerrenrut committed:
Review actions for #6417

Extended wx.SpinCtrl, the text is now selected when it gains focus.
Removed bounds checks in the onOk methods, used these values as limits
for spinCtrls

leonardder commented:
Would it be possible to do something to this new subclass in order to make NVDA indicate that this control contains a spin button when you tab to it? Now, there's no way for an end user to distinguish spin controls from regular text controls, other than just thinking "this is numerical input, so this might be a spin control"
You should also consider keeping the selection when someone changes the input using the arrow keys or the spin button, I think. Otherwise, following will happen:

1. A spin control contains a selected number
2. The number is changed using the arrow keys
3. There's no selection any more
4. One decides to change the number using numerical input
5. The numerical input is appended to the current value, however the numerical input should still replace the current value imo

jcsteh commented on source/gui/nvdaControls.py:
```diff
@@ -0,0 +1,25 @@
+	This makes editing the values quicker.
+	"""
+	def __init__(self, parent, id=wx.ID_ANY, value=wx.EmptyString, pos=wx.DefaultPosition, size=wx.DefaultSize, style=wx.SP_ARROW_KEYS|wx.ALIGN_RIGHT, min=0, max=100, initial=0, name=wx.SpinCtrlNameStr):
+		""" initialiser - Takes the same parameter as a wx.SpinCtrl.
```
parameter -> parameters

feerrenrut commented:
@leonardder we have discussed this a little and for the following reasons we won't implement these changes for this issue:
- Firefox, Chrome and Microsoft Word don't select text when you press up/down arrow to increment/decrement. While there aren't solid "standards" here, we think it's reasonable to follow this pattern.
- There are several edge cases and usability concerns that stop this being quick to implement and make it easy to implement badly. Issues like double announcing the value e.g. "23 unselected 24 selected."
- It seems like a minor use case to increment/decrement the value, then want to replace it by typing a value. More likely might be to type an approximate value then adjust it.

In terms of improving the description of the control, the work required here would be valuable to help fix several issues. I don't think we can justify the cost of implementation for this issue alone. I will raise a new umbrella issue to group together examples of issues that could be addressed together.

feerrenrut committed:
review actions for #6417

feerrenrut committed:
remove unused import

feerrenrut committed:
Merge remote-tracking branch 'origin/master' into i6099-useWxSpinCtrl

Conflicts:
	source/gui/nvdaControls.py
	source/gui/settingsDialogs.py

leonardder commented:
Fixes #4529
